[flang] Consolidate copy-in/copy-out determination in evaluate framework

eugeneepshteyn opened the pull request:
New implementation of `MayNeedCopy()` is used to consolidate copy-in/copy-out checks.

`IsAssumedShape()` and `IsAssumedRank()` were simplified and are both now in `Fortran::semantics` workspace.

`preparePresentUserCallActualArgument()` in lowering was modified to use `MayNeedCopyInOut()` 

Fixes https://github.com/llvm/llvm-project/issues/138471

eugeneepshteyn committed:
[flang][DRAFT] Copy-in/Copy-out determination

Plumbing/API for copy-in/copy-out

eugeneepshteyn committed:
Call DetermineCopyInOut() from lowering

eugeneepshteyn commented on flang/lib/Evaluate/call.cpp:
```diff
@@ -247,4 +248,19 @@ ProcedureRef::~ProcedureRef() {}
+  }
+  // TODO: at this point have dummy arguments as procInfo->dummyArguments
+  // and have actual arguments via arguments_
+}
```
Here go through the actual args and dummy args and determine copy-in and copy-out needs.

eugeneepshteyn committed:
clang-format

klausler reviewed (commented):
It might be cleaner to do the analysis at the time that the `ProcedureRef` is created by expression semantics rather than patching the IR later in lowering.  It could then be used in Semantics/check-call.cpp for all of the errors and warnings that depend on copy-in/copy-out instead of the ad-hoc analysis currently done that relies on contiguity &c.

eugeneepshteyn committed:
DetermineCopyInOut() is now called from ProcedureRef constructor

eugeneepshteyn committed:
clang-format

eugeneepshteyn committed:
Minor tweak

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
DetermineCopyInOut() is now called at ProcedureRef instantiation

eugeneepshteyn committed:
Very rough beginnings of argument handling in DetermineCopyInOut()

eugeneepshteyn committed:
More args handing in DetermineCopyInOut()

eugeneepshteyn committed:
clang-format

eugeneepshteyn committed:
clang-format

eugeneepshteyn committed:
clang-format

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Initial implementation of DetermineCopyInOutArgument() for implicit
interfaces.

Certain helper routines gained overloads to work with ActualArgument.

eugeneepshteyn committed:
Removed empty line

eugeneepshteyn committed:
clang-format

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-review

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Merge branch 'copy-inout-dev' into copy-inout-review

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Braces!

eugeneepshteyn committed:
clang-format

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Renamed copy-in/copy-out getter/setter functions

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Contiguity check

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
clang-format

eugeneepshteyn committed:
Continue filling out DetermineCopyInOutArgument(). Implemented IsExplicitShape(const Shape&)

eugeneepshteyn committed:
clang-format

eugeneepshteyn committed:
In DetermineCopyInOutArgument(), handle INTENT(IN) and INTENT(OUT)

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Support polymorphic-to-non-polymorphic case. Initial hookup into lowering, have LIT test failures

eugeneepshteyn committed:
clang-format

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
array check

eugeneepshteyn committed:
array check

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Changed IsExplicitShape() and restructured the checks

eugeneepshteyn committed:
Redid the integration to basically the old code but debug output with the new code

eugeneepshteyn committed:
Tweaks and debug output

eugeneepshteyn committed:
Switched to the new code. 5 failed tests left

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Fixed one issue with assumed rank arrays

eugeneepshteyn committed:
clang-format

eugeneepshteyn committed:
Expaned IsAssumedShape() checks and moved them to Fortran::evaluate namespace. In DetermineCopyInOutArgument() check for assumed shape and assumed rank

eugeneepshteyn committed:
clang-format

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Passes LIT tests

eugeneepshteyn committed:
Clean up in lowering to switch to the new checks

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Ignore rank support

eugeneepshteyn committed:
Polymorphic dummy with ignore rank

eugeneepshteyn committed:
clang-format (upstream)

eugeneepshteyn committed:
clang-format (upstream)

eugeneepshteyn committed:
clang-format (started using upstream)

eugeneepshteyn committed:
Added LIT test

eugeneepshteyn committed:
Implement copy-in/copy-out determination in Fortran::evaluate::MayNeedCopyInOut()

eugeneepshteyn committed:
clang-format

eugeneepshteyn committed:
Removed the old way to integrate copy-in/copy-out check

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-review

eugeneepshteyn committed:
Merge branch 'copy-inout-dev' into copy-inout-review

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Fixed the issue with user defined assignment

eugeneepshteyn committed:
clang-format

eugeneepshteyn committed:
Tweaked MayNeedCopyInOut() API

eugeneepshteyn committed:
Simplified IsAssumedRank() and IsAssumedShape()

eugeneepshteyn committed:
Moved IsAssumedRank() and IsAssumedShape() to namespace Fortran::semantics

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-review

eugeneepshteyn committed:
Merge branch 'copy-inout-dev' into copy-inout-review

eugeneepshteyn committed:
clang-format

eugeneepshteyn committed:
Merge branch 'copy-inout-dev' into copy-inout-review

vzakhari commented:
Thank you, Eugene! The lowering part looks good to me.

klausler commented on flang/include/flang/Evaluate/check-expression.h:
```diff
@@ -160,5 +163,31 @@ extern template bool IsErrorExpr(const Expr<SomeType> &);
+std::pair<bool, bool> MayNeedCopyInOut(const ActualArgument &,
+    const characteristics::DummyArgument &, FoldingContext &);
+
+inline std::pair<bool, bool> MayNeedCopyInOut(const ActualArgument *actual,
```
This API is a bit messy and it makes both the implementation and the client usage more complicated than they need to be.  You could do this with one API like this:

```
bool MayNeedCopy(const ActualArgument *, const characteristics::DummyArgument *, FoldingContext &, bool forCopyOut)
```

This would dodge the return of two Booleans that in this current PR prevents the use of quick `return`s for cases like `INTENT(IN)`/`INTENT(OUT)`, where you have lambdas to handle those cases instead.  It would also avoid having two distinct implementations for implicit and explicit procedure interfaces.  In the client code in lowering, the copy-out determination wouldn't be wasted work in the case of a vector subscript.

eugeneepshteyn replied:
I updated the API.

eugeneepshteyn committed:
WIP: new API MayNeedCopy()

eugeneepshteyn committed:
WIP: new API MayNeedCopy()

eugeneepshteyn committed:
clang-format

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Continue implementation of new MayNeedCopy()

eugeneepshteyn committed:
Assumed type check for contiguity

eugeneepshteyn committed:
Better void* standing for old implementation

eugeneepshteyn committed:
clang-format

eugeneepshteyn committed:
Tweaks to MayNeedCopyIn() to bring it closer to the old implementation

eugeneepshteyn committed:
Removed the old implementation (MayNeedCopyInOut) and switched to the new implementation

eugeneepshteyn committed:
Simplified MayNeedCopyOut()

eugeneepshteyn committed:
Enhanced flang/test/Lower/force-temp.f90 to cover more cases

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-review

eugeneepshteyn committed:
Merge branch 'copy-inout-dev' into copy-inout-review

eugeneepshteyn committed:
Fixed bad merge

eugeneepshteyn commented:
I updated the implementation to use new API and also added more checks to the new LIT test.

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Yet another refactoring and logic update: use common class CopyInOutExplicitInterface. Reduce the number of helper functions to two, one for copy-in and one for copy-out

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
clang-format

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
debugging (checking the old code again)

eugeneepshteyn committed:
Fixed build issue with the old code

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Different copy-in/copy-out flag configuration

eugeneepshteyn committed:
Removed the old code

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
updated comment

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-review

eugeneepshteyn committed:
Merge branch 'copy-inout-dev' into copy-inout-review

klausler commented on flang/lib/Evaluate/check-expression.cpp:
```diff
@@ -1434,4 +1446,212 @@ std::optional<parser::Message> CheckStatementFunction(
   return StmtFunctionChecker{sf, context}(expr);
 }
 
+// Helper class for cheching differences between actual and dummy arguments
```
"checking"

klausler commented on flang/lib/Evaluate/check-expression.cpp:
```diff
@@ -1434,4 +1446,212 @@ std::optional<parser::Message> CheckStatementFunction(
+        (dummyTreatAsArray && !dummyIsPolymorphic) || dummyIsVoidStar ||
+        dummyObj_.attrs.test(
+            characteristics::DummyDataObject::Attr::Contiguous)};
+    if (!actualTreatAsContiguous && dummyNeedsContiguity) {
```
Last four lines could be a single `return` statement.

klausler commented on flang/lib/Evaluate/check-expression.cpp:
```diff
@@ -1434,4 +1446,212 @@ std::optional<parser::Message> CheckStatementFunction(
+  const characteristics::DummyDataObject &dummyObj_;
+};
+
+static bool MayNeedCopyIn(FoldingContext &fc, const ActualArgument &actual,
```
Can you combine this function with the next one?  It might be clearer and more maintainable.

eugeneepshteyn replied:
Ok, simplified it further

klausler reviewed (commented):
Really looking good now!

eugeneepshteyn commented on flang/include/flang/Evaluate/call.h:
```diff
@@ -110,6 +110,7 @@ class ActualArgument {
 
   std::optional<DynamicType> GetType() const;
   int Rank() const;
+  bool IsArray() const;
```
> In the terminology of the Fortran standard, "assumed-rank" dummy arguments are actually not arrays.

To avoid confusion, I'll move the checks elsewhere.

eugeneepshteyn committed:
Code review feedback

eugeneepshteyn committed:
clang-format

eugeneepshteyn committed:
force-temp.f90: add test for INTENT(OUT) dummy argument

eugeneepshteyn committed:
MayNeedCopy() simplification

eugeneepshteyn committed:
Further simplification of MayNeedCopy()

eugeneepshteyn committed:
Removed the old code

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-review

eugeneepshteyn committed:
Merge branch 'copy-inout-dev' into copy-inout-review

eugeneepshteyn committed:
Removed DummyDataObject::IsArray() (no longer used)

eugeneepshteyn committed:
Merge branch 'copy-inout-dev' into copy-inout-review

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

klausler commented on flang/lib/Evaluate/check-expression.cpp:
```diff
@@ -1434,4 +1446,177 @@ std::optional<parser::Message> CheckStatementFunction(
+      //           non-contiguous.
+      // Copy-out: vector subscripts could refer to duplicate elements, can't
+      //           copy out.
+      return forCopyOut ? !HasVectorSubscript(*actual) : true;
```
`!(forCopyOut && HasVectorSubscript(...))` would be clearer, I think.

klausler reviewed (approved):
Looks great; thank you.

eugeneepshteyn committed:
Code review feedback

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-dev

eugeneepshteyn committed:
Merge branch 'llvm:main' into copy-inout-review

eugeneepshteyn committed:
Merge branch 'copy-inout-dev' into copy-inout-review
